### PR TITLE
Add jsonschema entrypoint to CLI

### PR DESCRIPTION
**Pull request by vankesteren**
- Overhauled argument parsing for CLI
- Better error message for top-level CLI call
- Automatic dispatch to different entrypoints in `__main__.py` file

### CODE REVIEW

**Comment by vankesteren** [comment]
Part of #151

File: metasynth/__main__.py [hunk 1]
```diff
@@ -1,56 +1,86 @@
...
+try:  # Python < 3.10 (backport)
+    from importlib_metadata import entry_points
+except ImportError:
+    from importlib.metadata import entry_points  # type: ignore [assignment]
+
 from metasynth import MetaFrame
+from metasynth._version import version_tuple
+from metasynth.validation import create_schema
+
+SEMVER = f"{version_tuple[0]}.{version_tuple[1]}.{version_tuple[2]}"
```
Comment (qubixes) [inline metasynth/__main__.py:17]: Why not use the actual version?

File: metasynth/__main__.py [hunk 2]
```diff
@@ -1,56 +1,86 @@
...
+    -h, --help    - display this help file and exit
+"""
+
+ENTRYPOINTS = ["synthesize", "jsonschema"]
+
 
+def main() -> None:
+    """CLI pointing to different entrypoints."""
+    # show help by default, else consume first argument
+    subcommand = "--help" if len(sys.argv) < 2 else sys.argv.pop(1)
```
Comment (qubixes) [inline metasynth/__main__.py:39]: It looks like argparse has functionality for sub-commands natively. Or is there a reason to do it manually?

https://docs.python.org/3/library/argparse.html#sub-commands

File: metasynth/__main__.py [hunk 3]
```diff
@@ -80,13 +110,55 @@ def main():
...
+        "plugins",
+        help="Plugins to include in the generated schema (default none)",
+        nargs="*",
+    )
+
+    parser.add_argument(
+        "-a", "--avail",
+        help="display available modules and quit",
+        action="store_true",
+    )
```
Comment (qubixes) [inline metasynth/__main__.py:139]: I think this should be `-l/--list`, which seems to be the standard.

File: metasynth/__main__.py [hunk 4]
```diff
@@ -80,13 +110,55 @@ def main():
...
+def jsonschema() -> None:
+    """Program to generate json schema from dist providers."""
+    parser = argparse.ArgumentParser(
+        prog="metasynth jsonschema",
+        description="Create Generative Metadata Format jsonschema and print to console.",
+    )
+
+    parser.add_argument(
+        "plugins",
+        help="Plugins to include in the generated schema (default none)",
```
Comment (qubixes) [inline metasynth/__main__.py:131]: This should be default builtin, no?

File: metasynth/__main__.py [hunk 5]
```diff
@@ -80,13 +110,55 @@ def main():
...
+    )
+
+    # parse the args without the subcommand
+    args, _ = parser.parse_known_args()
+
+    # deduplicated list of plugins for schema
+    plugins_avail = {entry.name for entry in entry_points(group="metasynth.distribution_provider")}
+
+    if args.avail:
+        _ = [print(a) for a in plugins_avail if a != "builtin"]
```
Comment (qubixes) [inline metasynth/__main__.py:148]: I think it's clearer if it's just a for-loop.

File: metasynth/__main__.py [hunk 6]
```diff
@@ -68,7 +98,7 @@ def main():
     if args.num_rows is not None:
         data_frame = meta_frame.synthesize(args.num_rows)
     else:
-        data_frame = meta_frame.synthesize(meta_frame.n_rows)
+        data_frame = meta_frame.synthesize(meta_frame.n_rows)  # type: ignore [arg-type]
```
Comment (qubixes) [inline metasynth/__main__.py:101]: Why is here an ignored type check? With the latest version, should it not accept None values?

**Review by qubixes — CHANGES_REQUESTED** [review]
That's a very nice addition! I'm still debating whether it should be `metasynth jsonschema`, or `metasynth schema`, maybe the second also covers it? I didn't know myself about the sub parsers, but I think it would fit well in our project to use them!

File: metasynth/__main__.py (hunk 2 above)
Comment (vankesteren) [reply metasynth/__main__.py:36]: won't fix in this PR if that's ok. Could not get the help file to display how I want. We can fix it if we want to add a new entrypoint or this file becomes more complex.

File: metasynth/__main__.py [hunk 7]
```diff
@@ -80,13 +108,58 @@ def main():
...
+        action="store_true",
+    )
+
+    # parse the args without the subcommand
+    args, _ = parser.parse_known_args()
+
+    # deduplicated list of plugins for schema
+    plugins_avail = {entry.name for entry in entry_points(group="metasynth.distribution_provider")}
+
+    if args.list:
```
Comment (qubixes) [inline metasynth/__main__.py:146]: I think for now it's okay to make "builtin" special, but I think if other packages are coming up, we might want to not make it special.

**Review by qubixes — APPROVED** [review]
Looks good!!

**Pull request merged**